feat : 내 스케줄/고정 스케줄 관리 기능

High-Quality-Coffee opened the pull request:
feat : 내 스케줄/고정 스케줄 관리 기능 추가
목적 :
- 내 스케줄 관리 기능 추가
- 고정 스케줄 관리 기능 추가

목표:
- 내 스케줄 CRUD
- 고정 스케줄 CRUD
- 게이트웨이 서버에 스케줄 서비스 등록

달성도 :
- 구현 완료

기타 :
- 코드 리뷰 필요

High-Quality-Coffee committed:
feat : 내 스케줄/고정 스케줄 관리 기능 추가

목적 :
- 내 스케줄 관리 기능 추가
- 고정 스케줄 관리 기능 추가

목표:
- 내 스케줄 CRUD
- 고정 스케줄 CRUD

달성도 :
- 구현 완료

기타 :
- 코드 리뷰 필요
- 게이트웨이 서버에 schedule-service 등록 필요

High-Quality-Coffee committed:
refactor : apigateway-service에 schedule-service 등록

목적 :
- 게이트웨이를 거쳐서 schedule-service를 사용할 수 있도록 하기 위해서 schedule-service 를 등록

목표 : application.yml 에 schedule-service 등록

kim946509 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/dto/MyFixScheduleInputDTO.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.time.LocalTime;
+
+@Data
```
저희  깃허브 Wiki 코드 컨벤션 보시면 무분별한 수정을 막고자 @Setter 사용을 지양하고 있습니다.
@Data안에 Setter가 포함되어 있어서 이것도 사용 지양하면 좋을것 같습니다.

kim946509 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/controller/MyScheduleController.java:
```diff
@@ -0,0 +1,62 @@
+     * @RequestHeader username, role : 유저 정보
+     * @param myScheduleInputDTO : MySchedule 생성 DTO
+     * @return : ResponseEntity<Map<String, String>>
+     **/
```
Method에 대한 주석은 @MethodDescription 어노테이션을  사용하면 좋을것 같습니다!

kim946509 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/global/feign/ScheduleServiceClient.java:
```diff
@@ -0,0 +1,14 @@
+@FeignClient(name = "user-service")
+public interface ScheduleServiceClient {
+    @GetMapping("/user-service/api/user/inquiry")
+    ResponseEntity<ApiResponseData<Long>> getUserIdByUsername(@RequestParam String username);
```
앗 혹시 User-service쪽에서 ApiResponseData가 아니라 예외가 발생해서 ApiResponseError객체가 전달되면 어떻게 될까요? 저도 이부분 고민중이여서요.

kim946509 replied:
그리고 혹시 username으로 본인을 검사한 이유가 있을까요? 제 생각에는 이 상황에서는 닉네임이 더 적합해보여서요..!

kim946509 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/service/MyFixScheduleService.java:
```diff
@@ -0,0 +1,95 @@
+        MyFixScheduleEntity myFixScheduleEntity = myFixScheduleRepository.findById(scheduleId).orElseThrow(()->new BaseException("삭제할 시간표를 찾을 수 없습니다."));
+        //수정하려는 시간표가 내 시간표인지 확인
+        if(!userId.equals(myFixScheduleEntity.getUserId())){
+            throw new BaseException("다른 사람의 시간표를 수정할 수 없습니다.");
```
모두 BaseException으로 예외가 발생하고 코드가 전달되지 않으면, 아마 기본 code가 전달될것 같은데 별도의 커스텀 예외를 만들거나, code를 지정해주는건  어떨까요?

kim946509 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/ScheduleServiceApplication.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.cloud.openfeign.EnableFeignClients;
+
+@SpringBootApplication
+@EnableFeignClients
```
별도의 Config 패키지에  추가하는게 좋을것 같아요!

kim946509 commented:
고생하셨습니다!!

ryudain05 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/service/MyFixScheduleService.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class MyFixScheduleService {
```
해당 서비스는 조회 메소드를 제외하고는  모두 @Transactional 을 붙여야할 것 같다고 생각합니다. 클래스단에 @Transactional 를 붙이고, 조회 메소드에서는 @Transactional(readOnly = true) 를 붙이는 게 어떨까요???

```

@Service
@RequiredArgsConstructor
@Transactional
public class MyFixScheduleService {

@Transactional(readOnly = true)
public List<MyFixScheduleOutputDTO> getMyFixSchedule(Long userId){
```

ryudain05 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/service/MyFixScheduleService.java:
```diff
@@ -0,0 +1,95 @@
+    //고정 스케줄 삭제
+    @Transactional
+    public void deleteMyFixSchedule(Long userId, Long scheduleId){
+        if(userId == null){
+            throw new BaseException("유저의 시간표를 찾을 수 없습니다.");
+        }
+
+        if(scheduleId == null){
+            throw new BaseException("해당 시간표는 존재하지 않습니다.");
+        }
```
이렇게 중복되는 부분들은 내부 메소드로 빼도 될 것 같습니다 !

ryudain05 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/service/MyScheduleService.java:
```diff
@@ -0,0 +1,79 @@
+
+@Service
+@RequiredArgsConstructor
+public class MyScheduleService {
```
해당 서비스에 @Transactional 이 ,, 없는 듯 합니다 !

ryudain05 commented on src/backend/schedule-service/src/main/java/site/danjam/mate/schedule_service/entity/MyFixScheduleEntity.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.time.LocalTime;
+
+@Entity
+@Data
+@Table(name = "myfixschedule")
+public class MyFixScheduleEntity {
```
Entity에 기본 생성자가 없는 듯 합니다 !! Jpa Proxy 조회 한번 보시면 좋을 거 같아요 ㅎㅎ

ryudain05 reviewed (approved):
수고하셨습니다 !!!!!!!!!!!!!!! 🙆🏻‍♀️

High-Quality-Coffee committed:
refactor : 리뷰 바탕으로 전반적인 코드 수정

목적 :
- pr 코드 리뷰 바탕으로 코드 수정

기타 :
- @transactional 동작방식 파악하기
- JPA proxy 동작방식 확인하기
- 스케줄 서비스 테스트 필요

High-Quality-Coffee committed:
refactor : 에러 처리 추가

목적 : 해당 스케줄이 존재하지 않는다는 에러 처리
